lambdas/shared/t4_lambda_shared/preview.py

```diff
@@ -129,81 +129,80 @@ def extract_fcs(file_, as_html=True):
-        size_guess = num_rows_guess * meta.num_columns * AVG_PARQUET_CELL_BYTES
-        if skip_rows or (size_guess > get_available_memory()):
-            # minimal dataframe with all columns and one row
-            dataframe = pandas.DataFrame(columns=meta.schema.names)
-            info['warnings'] = 'Large file: skipped rows to conserve memory, only showing column names'
-        else:
-            dataframe = pf.read_row_group(0)[:MAX_PREVIEW_ROWS].to_pandas()
-    # sometimes there are neither rows nor row_groups, just columns
-    # therefore we do not call read_row_group because (with 0 row_groups)
-    # it would barf
-    else:
-        # :0 is a safety valve but there really should be no rows in this case
-        dataframe = pf.read()[:0].to_pandas()
-    if as_html:
-        body = remove_pandas_footer(dataframe._repr_html_())  # pylint: disable=protected-access
+
+    available = get_available_memory()
+    iter_batches = None
+    # 10MB heuristic; should never happen, e.g. with current default of 512MB
+    if (available < 10E6) or skip_rows:
+        logger_.warning("Insufficient memory to index parquet file: %s", info)
+        info['warnings'] = "Skipped rows; insufficient memory"
+    elif meta.num_rows and meta.num_row_groups:
+        iter_batches = pf.iter_batches(batch_size=128, row_groups=[0])
     else:
+        logger_.warning("Parquet file with no rows: %s", info)
+    if iter_batches:
         buffer = []
         size = 0
         done = False
-        for _, row in dataframe.iterrows():
-            for column in row.astype(str):
-                encoded = column.encode()
-                # +1 for \t
-                encoded_size = len(encoded) + 1
-                if (size + encoded_size) < max_bytes:
-                    buffer.append(encoded)
-                    buffer.append(b"\t")
-                    size += encoded_size
-                else:
-                    done = True
-                    break
-            buffer.append(b"\n")
-            size += 1
+        for batch in iter_batches:
+            df = batch.to_pandas()
+            if as_html:
+                body = remove_pandas_footer(df._repr_html_())
+                return body, info
+            for _, row in df.iterrows():
+                for column in row.astype(str):
+                    encoded = column.encode()
+                    encoded_size = len(encoded) + 1  # +1 for \t
+                    if (size + encoded_size) < max_bytes:
+                        buffer.append(encoded)
+                        buffer.append(b"\t")
+                        size += encoded_size
+                    else:
+                        done = True
+                        break
+                buffer.append(b"\n")
+                size += 1
             if done:
                 break
         body = b"".join(buffer).decode()
+    elif as_html:
+        body = remove_pandas_footer(pandas.DataFrame(columns=meta.schema.names)._repr_html_())
```

Comment:
Do we need the 'Large file: skipped rows to conserve memory, only showing column names' here?

Reply:
I don't think so because if this _function_ skips rows we triggered the condition on line 172. So if we reached this far we know there are no batches and that can only be because the file itself had no rows.